game/script-farewells.rpy

```diff
@@ -2,42 +2,1223 @@ default persistent._farewell_database = dict()
 
 init python in farewells:
     import random
+    import store
+
     FAREWELL_MAP = dict()
 
     def select_farewell():
         """
-        Picks a random greeting
+        Picks a random farewell, accounting for affinity
+        If the player has already been asked to stay by Natsuki, a farewell without the option
+        to stay will be selected
+        """
+        # Get the farewells the current affinity allows for the player
+        #TODO: Generalized filter function
+        farewell_pool = filter(lambda x: x.evaluate_affinity_range(), FAREWELL_MAP.values())
+
+        # Filter any time-sensitive topics, if the player has been around 30 minutes or longer
+        if divmod((store.utils.get_current_session_length()).total_seconds(), 60)[0] > 30:
+            farewell_pool = filter(
+                lambda x: store.Topic.get_topic_has_additional_property_with_value(x, "is_time_sensitive", False),
+                farewell_pool
+            )
+
+        # If Natsuki has already asked her player to stay, filter any topics that would let her ask again
+        if store.jn_globals.player_already_stayed_on_farewell:
+            farewell_pool = filter(
+                lambda x: store.Topic.get_topic_has_additional_property_with_value(x, "has_stay_option", False),
+                farewell_pool
+            )
+
+        # Finally return a random farewell from the remaining pool
+        return random.choice(farewell_pool).label
+
+    def try_trust_dialogue():
+        """
+        Coinflip decision on whether to additionally call trust-based dialogue on farewell
         """
-        #TODO: ME (for now this just returns a random)
-        return random.choice(FAREWELL_MAP.keys())
+        if random.choice([True, False]):
+            renpy.call_in_new_context("farewell_extra_trust")
+
+    def get_time_in_session_descriptor():
+        """
+        Get a descriptor based on the number of minutes the player has spent in the session, up to 30 minutes
+
+        OUT:
+            Brief descriptor relating to the number of minutes spent in the session
+        """
+        minutes_in_session = divmod((store.utils.get_current_session_length()).total_seconds(), 60)[0]
```

Comment:
Looking at this, is there any reason to use `divmod` here?

You save only the first item in the tuple, so it seems like it'd be more worthwhile to just use `total_seconds() / 60`

Reply:
That's actually a good question; I think I went with divmod here originally as some of the descriptor code used to refer to exact matches rather than gt/et or lt/et. I'll swap it out as it should change functionality